cmd/lakectl/cmd/retry_client_test.go

```diff
@@ -0,0 +1,236 @@
+			expectedShouldRetry: true,
+			expectedError:       "",
+		},
+		{
+			name: "HTTP Status - 503 Service Unavailable",
+			getTestContext: func() context.Context {
+				return context.Background()
+			},
+			resp: &http.Response{
+				StatusCode: http.StatusServiceUnavailable,
+			},
+			err:                 nil,
+			expectedShouldRetry: true,
+			expectedError:       "",
+		},
+		{
+			name: "HTTP Status - 401 Unauthorized",
+			getTestContext: func() context.Context {
+				return context.Background()
+			},
+			resp: &http.Response{
+				StatusCode: http.StatusUnauthorized,
+			},
+			err:                 nil,
+			expectedShouldRetry: false,
+			expectedError:       "",
+		},
+		{
+			name: "HTTP Status - 404 Not Found",
+			getTestContext: func() context.Context {
+				return context.Background()
+			},
+			resp: &http.Response{
+				StatusCode: http.StatusNotFound,
+			},
+			err:                 nil,
+			expectedShouldRetry: false,
+			expectedError:       "",
+		},
+		{
+			name: "HTTP Status - 200 Ok",
+			getTestContext: func() context.Context {
+				return context.Background()
+			},
+			resp: &http.Response{
+				StatusCode: http.StatusOK,
+			},
+			err:                 nil,
+			expectedShouldRetry: false,
+			expectedError:       "",
+		},
+		{
+			name: "HTTP Status - 201 Created",
+			getTestContext: func() context.Context {
+				return context.Background()
+			},
+			resp: &http.Response{
+				StatusCode: http.StatusCreated,
+			},
+			err:                 nil,
+			expectedShouldRetry: false,
+			expectedError:       "",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			// if tc.err != nil {
+			// 	errStr := tc.err.Error()
+			// 	fmt.Sprintf("%s", errStr)
+			// }
+			shouldRetry, err := lakectlRetryPolicy(tc.getTestContext(), tc.resp, tc.err)
+			require.Equal(t, tc.expectedShouldRetry, shouldRetry)
+			if tc.expectedError != "" {
+				require.EqualError(t, err, tc.expectedError)
+			}
+		})
+	}
+}
+
+func TestRetryHTTPClient(t *testing.T) {
+	retryCount := -1
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.MaxIdleConnsPerHost = DefaultMaxIdleConnsPerHost
+	retryClient := NewRetryClient(10, 10*time.Millisecond, 30*time.Millisecond, transport)
+
+	httpCode := int64(http.StatusTooManyRequests)
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		retryCount++
+		w.WriteHeader(int(atomic.LoadInt64(&httpCode)))
+	}))
+	defer server.Close()
+
+	req, err := http.NewRequest(http.MethodGet, server.URL, nil)
+	if err != nil {
+		t.Fatal("failed to create request")
+	}
+
+	var resp *http.Response
+	doneCh := make(chan struct{})
+	errCh := make(chan error, 1)
+	go func() {
+		defer close(doneCh)
+		defer close(errCh)
+		var err error
+		resp, err = retryClient.Do(req)
+		errCh <- err
+	}()
+
+	select {
+	case <-doneCh:
+		t.Fatal("should retry on 429 http status")
+	case <-time.After(200 * time.Millisecond):
```

Comment:
This test appears to depend on a race, so it will be very flaky on CI.

Consider structuring the test differently.  For instance, the server could send errors that tell the client to retry 3 times and then succeed.  Then you would know exactly what happened.

Reply:
Clarification: I'm not saying we shouldn't test the surface. I'm saying that testing `lakectl` with a mock `lakeFS` server would sufficiently cover this, in addition to the coverage provided by the tests in the package repo.